packages/core/src/manifest.ts

```diff
@@ -38,27 +41,103 @@ function defaultManifest(): ManifestData {
   };
 }
 
-const manifestDir = '.openzeppelin';
+const MANIFEST_DEFAULT_DIR = '.openzeppelin';
+const MANIFEST_TEMP_DIR = 'openzeppelin-upgrades';
+
+async function getDevInstanceMetadata(
+  provider: EthereumProvider,
+  chainId: number,
+): Promise<DevInstanceMetadata | undefined> {
+  let hardhatMetadata;
+
+  try {
+    hardhatMetadata = await getHardhatMetadata(provider);
+  } catch (e: unknown) {
+    return undefined;
+  }
+
+  if (hardhatMetadata.chainId !== chainId) {
+    throw new Error(
+      `Broken invariant: Hardhat metadata's chainId ${hardhatMetadata.chainId} does not match eth_chainId ${chainId}`,
+    );
+  }
+
+  return {
+    networkName: 'hardhat',
+    instanceId: hardhatMetadata.instanceId,
+    forkedNetwork: hardhatMetadata.forkedNetwork,
+  };
+}
+
+function getSuffix(chainId: number, devInstanceMetadata?: DevInstanceMetadata) {
+  if (devInstanceMetadata !== undefined) {
+    return `${chainId}-${devInstanceMetadata.instanceId}`;
+  } else {
+    return chainId;
+  }
+}
+
+interface DevInstanceMetadata {
+  networkName: string;
+  instanceId: string;
+  forkedNetwork?: {
+    // The chainId of the network that is being forked
+    chainId: number;
+  };
+}
 
 export class Manifest {
   readonly chainId: number;
   readonly file: string;
-  private readonly fallbackFile: string;
+  readonly fallbackFile: string;
+
+  private readonly devInstanceMetadata?: DevInstanceMetadata;
+  private readonly devDir?: string;
+  private readonly parent?: Manifest;
 
   private locked = false;
 
   static async forNetwork(provider: EthereumProvider): Promise<Manifest> {
-    return new Manifest(await getChainId(provider));
+    const chainId = await getChainId(provider);
+    const devInstanceMetadata = await getDevInstanceMetadata(provider, chainId);
+    if (devInstanceMetadata !== undefined) {
+      return new Manifest(chainId, devInstanceMetadata, os.tmpdir());
+    }
+    return new Manifest(chainId);
   }
 
-  constructor(chainId: number) {
+  constructor(chainId: number, devInstanceMetadata?: DevInstanceMetadata, osTmpDir?: string) {
     this.chainId = chainId;
+    this.devInstanceMetadata = devInstanceMetadata;
+
+    let forkedChainId = undefined;
```

Comment:
This variable is not needed outside the `if` statement as far as I can tell.

Reply:
Changed it save the chain id suffix (including instance id if applicable) as an instance variable, since that is needed to determine the lock file name.